File: backend/app/schemas/analytics.py
from datetime import date

from pydantic import BaseModel, Field


class AnalyticsStrategyBreakdown(BaseModel):
    strategy_name: str
    trades_count: int
    realized_pnl: float
    win_rate: float | None = None
    profit_factor: float | None = None
    average_r: float | None = None


class AnalyticsExecutionQuality(BaseModel):
    fills_count: int
    matched_fills_count: int
    review_needed_fills_count: int
    planned_entry_count: int
    average_entry_drag_r: float | None = None
    average_entry_slippage_pct: float | None = None
    planned_exit_count: int
    average_exit_drag_r: float | None = None


class AnalyticsUnrealizedPosition(BaseModel):
    position_id: str
    symbol_id: str
    strategy_name: str | None = None
    quantity: float
    entry_price: float | None = None
    mark_price: float | None = None
    mark_ts: str | None = None
    unrealized_pnl: float
    source: str


class AnalyticsOverviewResponse(BaseModel):
    from_date: date
    to_date: date
    equity: float
    total_pnl: float
    realized_pnl: float
    unrealized_pnl: float
    win_rate: float
    profit_factor: float | None
    expectancy_r: float
    average_r: float
    max_drawdown_pct: float
    current_drawdown_pct: float
    trades_count: int
    open_risk_pct: float
    open_positions_count: int = 0
    closed_positions_count: int = 0
    unrealized_positions: list[AnalyticsUnrealizedPosition] = Field(default_factory=list)
    strategy_breakdown: list[AnalyticsStrategyBreakdown] = Field(default_factory=list)
    execution_quality: AnalyticsExecutionQuality = Field(
        default_factory=lambda: AnalyticsExecutionQuality(
            fills_count=0,
            matched_fills_count=0,
            review_needed_fills_count=0,
            planned_entry_count=0,
            planned_exit_count=0,
        )
    )

File: frontend/src/lib/api/types/analytics.ts
export type AnalyticsStrategyBreakdown = {
  strategy_name: string;
  trades_count: number;
  realized_pnl: number;
  win_rate: number | null;
  profit_factor: number | null;
  average_r: number | null;
};

export type AnalyticsExecutionQuality = {
  fills_count: number;
  matched_fills_count: number;
  review_needed_fills_count: number;
  planned_entry_count: number;
  average_entry_drag_r: number | null;
  average_entry_slippage_pct: number | null;
  planned_exit_count: number;
  average_exit_drag_r: number | null;
};

export type AnalyticsUnrealizedPosition = {
  position_id: string;
  symbol_id: string;
  strategy_name: string | null;
  quantity: number;
  entry_price: number | null;
  mark_price: number | null;
  mark_ts: string | null;
  unrealized_pnl: number;
  source: string;
};

export type AnalyticsOverview = {
  from_date: string;
  to_date: string;
  equity: number;
  total_pnl: number;
  realized_pnl: number;
  unrealized_pnl: number;
  win_rate: number;
  profit_factor: number | null;
  expectancy_r: number;
  average_r: number;
  max_drawdown_pct: number;
  current_drawdown_pct: number;
  trades_count: number;
  open_risk_pct: number;
  open_positions_count: number;
  closed_positions_count: number;
  unrealized_positions: AnalyticsUnrealizedPosition[];
  strategy_breakdown: AnalyticsStrategyBreakdown[];
  execution_quality: AnalyticsExecutionQuality;
};
